### PR TITLE
Bring in QED-it Tests

### PR DESCRIPTION
**Pull request by defuse**
This PR was originally submitted here: https://github.com/zcash-hackworks/sapling-crypto/pull/101

This pulls in a bunch of tests QED-it wrote during their audit that were spread across many branches in their [sapling-crypto-internal](https://github.com/QED-it/sapling-crypto-internal) repository. I've left their history intact so the best way to review it is to just look at the final diff, except for...

- [This commit](https://github.com/zcash/librustzcash/pull/93/commits/f6f5f586aaae5d652bcd082ea82d3e0de24ac60f), which includes changes outside of tests, and
- [This commit](https://github.com/zcash/librustzcash/pull/93/commits/335230448dec0c2ccd4aa230a053866b4ffedf9d), which is my attempt to fix the broken linear relation testing in that commit (the check in that commit effectively just checks that p1 != p2 again, not what it was intended to do).
- I deleted a [`find_group_hash` test](https://github.com/zcash/librustzcash/pull/93/commits/13bf8d5f0f9bb221cf5bd2a88e9c0b4a46d077ed) since it didn't look like it was doing anything useful, and I don't want merging to block on me figuring out how to implement that test well. Raise a separate issue and assign me if you would like some `find_group_hash` unit tests!

I did *not* bring in [this change to the circuit code](https://github.com/QED-it/sapling-crypto-internal/commit/31781e844fe976649c817ed3fc32c683aa383c76), because I don't understand it. I also did not bring in a bunch of changes to comments and variable renamings in [the aurel_comments branch](https://github.com/QED-it/sapling-crypto-internal/compare/aurel_comments) and [the sum_bug branch](https://github.com/QED-it/sapling-crypto-internal/compare/sum_bug), but I did bring in the new tests from those branches.

### CODE REVIEW

**Comment by defuse** [comment]
This needs some cleanup and improvements before it can be merged. In particular the group_hash tests are incomplete.

File: sapling-crypto/src/circuit/ecc.rs [hunk 1]
```diff
@@ -1208,4 +1208,79 @@ mod test {
...
+        assert_ne!(zero_x, should_not_be_zero.into_xy().0);
+        assert_ne!(zero_y, should_not_be_zero.into_xy().1);
+
+        let should_be_zero = should_not_be_zero.add(&g_small, params);
+        assert_eq!(zero_x, should_be_zero.into_xy().0);
+        assert_eq!(zero_y, should_be_zero.into_xy().1);
+
+        // take all the points from the script
+        // assert should be different than multiplying by cofactor, which is the solution
+        // is user input verified? https://github.com/zcash/librustzcash/blob/f5d2afb4eabac29b1b1cc860d66e45a5b48b4f88/src/rustzcash.rs#L299
```
Comment (defuse) [inline sapling-crypto/src/circuit/ecc.rs]: I think this is talking about how we're missing an explicit check that `ask != 0` in our code? Should I just delete these comments?

**Comment by defuse** [comment]
Uh oh I accidentally pushed the branch to this repo, deleted it, then rebased and now Travis is broken: https://travis-ci.org/zcash/librustzcash/builds/566534948?utm_source=github_status&utm_medium=notification

```
$ git clone --depth=50 --branch=qed-it-lrz https://github.com/zcash/librustzcash.git zcash/librustzcash
Cloning into 'zcash/librustzcash'...
warning: Could not find remote branch qed-it-lrz to clone.
fatal: Remote branch qed-it-lrz not found in upstream origin
```

**Comment by str4d** [comment]
Now that #104 has been merged, this PR needs to be refactored to account for `sapling-crypto` being split up. I can do this @defuse if you don't have time.

**Comment by str4d** [comment]
I've started rebasing this, and I've decided that I'm not going to preserve all history because there is a lot of cruft within the commits, and in particular there are some commits that are clearly mislabeled (the code they refer to appears in an adjacent commit, or spread across several commits). I will refactor retaining the intention of the changes, and as much authorship as I can.

**Comment by str4d** [comment]
Rebased on master, cleaned up some of the commits (so they match what they say on the tin), and formatted most of the commits. The PR will probably need to be rebased again once #113 lands.

**Comment by defuse** [comment]
Added a commit to make use of all the test vectors. That required increasing the number of Pedersen hash generators to 6. Someone should check that I did so correctly. Alternatively I can delete the test vectors that are too long.

File: zcash_primitives/src/jubjub/mod.rs [hunk 2]
```diff
@@ -220,7 +220,7 @@ impl JubjubBls12 {
         {
             let mut pedersen_hash_generators = vec![];
 
-            for m in 0..5 {
+            for m in 0..6 {
```
Comment (str4d) [inline zcash_primitives/src/jubjub/mod.rs]: With #111 merged into master, this shouldn't be necessary.

Comment (str4d) [reply zcash_primitives/src/jubjub/mod.rs]: I checked during the rebase, and this is indeed necessary for the given test vectors.

File: zcash_primitives/src/test_vectors/pedersen_hash_vectors.rs [hunk 3]
```diff
@@ -0,0 +1,716 @@
+//! Test vectors from https://github.com/zcash-hackworks/zcash-test-vectors/blob/master/sapling_pedersen.py
```
[the remark this answers is not in the document]
Comment (defuse) [reply zcash_primitives/src/test_vectors/pedersen_hash_vectors.rs]: I think they're made by https://github.com/QED-it/go-jubjub

File: zcash_primitives/src/jubjub/mod.rs [hunk 4]
```diff
@@ -427,6 +400,65 @@ impl JubjubBls12 {
...
+                if p1 == p2 {
+                    panic!("Duplicate generator!");
+                }
+                if p1 == &p2.negate() {
+                    panic!("Inverse generator!");
+                }
+            }
+
+            // check for a generator being the sum of any other two
+            for (j, p2) in pedersen_hash_generators.iter().enumerate() {
```
Comment (Eirik0) [inline zcash_primitives/src/jubjub/mod.rs]: I experimented with trying to combine this loop with the one above, but wasn't able to get it to work properly. If this function were performance critical, we could first add all of the items to a HashSet, and then check if combinations of p1+p2 were in that set.

**Pull request merged**